src/Storages/MergeTree/MergeTreeMutationEntry.cpp

```diff
@@ -45,21 +48,35 @@ UInt64 MergeTreeMutationEntry::parseFileName(const String & file_name_)
     throw Exception(ErrorCodes::BAD_ARGUMENTS, "Cannot parse mutation version from file name, expected 'mutation_<UInt64>.txt', got '{}'", file_name_);
 }
 
-MergeTreeMutationEntry::MergeTreeMutationEntry(MutationCommands commands_, DiskPtr disk_, const String & path_prefix_, UInt64 tmp_number,
-                                               const TransactionID & tid_, const WriteSettings & settings)
+MergeTreeMutationEntry::MergeTreeMutationEntry(
+    MutationCommands commands_,
+    DiskPtr disk_,
+    const String & path_prefix_,
+    UInt64 tmp_number,
+    const std::optional<String> & partition_id_,
+    const TransactionID & tid_,
+    const WriteSettings & settings
+)
     : create_time(time(nullptr))
     , commands(std::move(commands_))
     , disk(std::move(disk_))
     , path_prefix(path_prefix_)
     , file_name("tmp_mutation_" + toString(tmp_number) + ".txt")
     , is_temp(true)
+    , partition_id(partition_id_)
     , tid(tid_)
 {
     try
     {
         auto out = disk->writeFile(std::filesystem::path(path_prefix) / file_name, DBMS_DEFAULT_BUFFER_SIZE, WriteMode::Rewrite, settings);
         *out << "format version: 1\n"
             << "create time: " << LocalDateTime(create_time) << "\n";
+        if (partition_id)
+        {
+            *out << "partition id: ";
+            writeEscapedString(*partition_id, *out);
+            *out << "\n";
+        }
```

Comment:
Do we really need to serialize `partition_id` here? Seems like we already have partition id inside `commands`

Reply:
We don't need serialization of it specifically, but we need a way to distinguish old storage-wide mutations from new. If we don't change format, we will try to recalculate `paritition_id` for older and possibly unfinished partition-specific mutations, skip unrelated partitions and stick on merges when in the same partition subset of parts is mutated beyond correct version, or stick being unable to mutate another subset beyond it because there is no related mutation for that. I think this serialization is a neat way to avoid that.

Reply:
I mean that we have different mutations:
1) with `partition_id` and `IN PARTITION` expression, which are new partition-specific mutations — these shall create partition-specific changes;
2) without `partition_id` and `IN PARTITION` expression, which are old or new storage-wide mutations — these create storage-wide changes;
3) without `partition_id` but with `IN PARTITION` expression, which are old partition-specific mutations. They were supposed to make storage-wide changes earlier (before this patch) and may be partially finished on some parts, so we must continue working the same way on them. If we remove these guys from remaining really unaffected parts (`parts_to_do`), we would have to rollback unaffected parts which already mutated to that version, or we will have finished mutation and different versions in a partition which will prevent merges to happen until new mutation (in that partition or whole storage) appears.

Generally, I think this is a good thing to have partition set to be calculated at the moment when the mutation is created.

Reply:
I understand your point, but I'm not sure about

> we must continue working the same way on them. If we remove these guys from remaining really unaffected parts (parts_to_do), we would have to rollback unaffected parts which already mutated to that version

I agree that it's undesirable to have some parts with data version greater than expected version for the partition. But AFAIK it will not break anything. 

> will prevent merges to happen until new mutation (in that partition or whole storage) appears

It will prevent only merges of parts with different data versions, seems like it's not critical. Also it probably can be addressed by taking `last_mutation_by_partition` into account in `getCurrentMutationVersion`.

I don't like serializing partition id for two reasons:
 - It duplicates partition id that we already have in `IN PARTITION` clause (and we already serialize it).
 - It unconditionally changes format of mutation entries without any settings. It will make downgrade impossible once you run `IN PARTITION` mutation. `IN PARTITION` is not a new feature, so this change is backward incompatible. We can introduce a setting like `merge_tree_mutation_entry_version` or `serialize_partition_id_for_in_partition_mutations` for compatibility, but I don't like this option. It's better to avoid changing format, and seems like it's possible to avoid it.